src/ucp/wireup/wireup_cm.c

```diff
@@ -733,21 +737,28 @@ static void ucp_cm_server_connect_cb(uct_ep_h ep, void *arg,
     uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
     ucp_lane_index_t cm_lane;
 
-    if (status != UCS_OK) {
-        /* if reject is arrived on server side, then UCT does something wrong */
-        ucs_assert(status != UCS_ERR_REJECTED);
+    if (status == UCS_OK) {
+        uct_worker_progress_register_safe(ucp_ep->worker->uct,
+                                          ucp_cm_server_connect_progress,
+                                          ucp_ep, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                          &prog_id);
+        ucp_worker_signal_internal(ucp_ep->worker);
+    } else if (status == UCS_ERR_CONNECTION_RESET) {
+        /* remote side initiated disconnect before local side has completed
+         * connection establishment, so:
+         * 1) establish connection to complete local flush
+         * 2) handle disconnect same way as close protocol */
+        uct_worker_progress_register_safe(ucp_ep->worker->uct,
+                                          ucp_cm_server_connect_progress,
+                                          ucp_ep, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                          &prog_id);
+        ucp_cm_disconnect_cb(ep, ucp_ep);
+    } else {
```

Comment:
the assert is removed because REJECT on the server is possible?

also, the doc of this function - should be "server is connected" ?

Reply:
> the assert is removed because REJECT on the server is possible?

by mistake, will return back, thanks!

> also, the doc of this function - should be "server is connected" ?

no, callback on server notifies state of client

Reply:
> no, callback on server notifies state of client

right, after client sent the 'established' message. by then the server is considered connected as well.

Reply:
right :)